rubix-spi/src/main/java/com/qubole/rubix/spi/fop/ObjectPoolPartition.java

```diff
@@ -50,115 +53,83 @@ public ObjectPoolPartition(ObjectPool<T> pool, PoolConfig config,
-    this.totalCount = 0;
+    this.totalCount = new AtomicInteger();
     this.log = new CustomLogger(name, host);
-    for (int i = 0; i < config.getMinSize(); i++) {
-      T object = objectFactory.create(host, socketTimeout, connectTimeout);
-      if (object != null) {
-        objectQueue.add(new Poolable<>(object, pool, host));
-        totalCount++;
+    this.takeSemaphore = new Semaphore(config.getMaxSize(), true);
+    try {
+      for (int i = 0; i < config.getMinSize(); i++) {
+          T object = objectFactory.create(host, socketTimeout, connectTimeout);
+          objectQueue.add(new Poolable<>(object, pool, host));
+          totalCount.incrementAndGet();
       }
     }
+    catch (Exception e) {
+      // skipping logging the exception as factories are already logging.
+    }
   }
 
   public void returnObject(Poolable<T> object)
   {
-    if (!objectFactory.validate(object.getObject())) {
-      log.debug(String.format("Invalid object...removing: %s ", object));
-      decreaseObject(object);
-      // Compensate for the removed object. Needed to prevent endless wait when in parallel a borrowObject is called
-      increaseObjects(1, false);
-      return;
-    }
+    try {
+      if (!objectFactory.validate(object.getObject())) {
+        log.debug(String.format("Invalid object...removing: %s ", object));
+        decreaseObject(object);
+        return;
+      }
 
-    log.debug(String.format("Returning object: %s to queue. Queue size: %d", object, objectQueue.size()));
-    if (!objectQueue.offer(object)) {
-      log.warn("Created more objects than configured. Created=" + totalCount + " QueueSize=" + objectQueue.size());
-      decreaseObject(object);
+      log.debug(String.format("Returning object: %s to queue. Queue size: %d", object, objectQueue.size()));
+      if (!objectQueue.offer(object)) {
+        log.warn("Created more objects than configured. Created=" + totalCount + " QueueSize=" + objectQueue.size());
+        decreaseObject(object);
+      }
+    }
+    finally {
+      takeSemaphore.release();
     }
   }
 
   public Poolable<T> getObject(boolean blocking)
   {
-    if (objectQueue.size() == 0) {
-      // increase objects and return one, it will return null if pool reaches max size or if object creation fails
-      Poolable<T> object = increaseObjects(this.config.getDelta(), true);
-
-      if (object != null) {
-        return object;
-      }
-
-      if (totalCount == 0) {
-        // Could not create objects, this is mostly due to connection timeouts hence no point blocking as there is not other producer of sockets
-        throw new RuntimeException("Could not add connections to pool");
-      }
-      // else wait for a connection to get free
-    }
-
-    Poolable<T> freeObject;
+    Poolable<T> object;
     try {
       if (blocking) {
-        freeObject = objectQueue.take();
+        takeSemaphore.acquire();
+        object = objectQueue.take();
```

Comment:
what's the reason for not calling `tryGetObject` here?

Reply:
We should. Or we should remove the code path for `blocking=true` case as it is not applicable for us and it is only complicating maintenance.

Reply:
tryGetObject call object.Queue.poll() which is non blocking. But, in case of blocking mode, though there is semaphore acquire blocking call, I just a called 'take' to be on safe side.

I guess better removing this blocking mode as we are not using it anymore in the code.